eta/core/models.py

```diff
@@ -0,0 +1,909 @@
+        `distutils.version.LooseVersion` intended for comparison operations.
+
+        Models with no version are given a version of 0.0.0.
+        '''
+        return LooseVersion(self.version or "0.0.0")
+
+    def get_path_in_dir(self, models_dir):
+        '''Gets the model path for the model in the given models directory.'''
+        return os.path.join(models_dir, self.filename)
+
+    def is_in_dir(self, models_dir):
+        '''Determines whether a copy of the model exists in the given models
+        directory.
+        '''
+        return os.path.isfile(self.get_path_in_dir(models_dir))
+
+    @staticmethod
+    def parse_name(name):
+        '''Parses the model name, returning the base name and the version,
+        if any.
+
+        Args:
+            name: the name of the model, which can have "@<ver>" appended to
+                refer to a specific version of the model
+
+        Returns:
+            base_name: the base name of the model
+            version: the version of the model, or None if no version was found
+
+        Raises:
+            ModelError: if the model name was invalid
+        '''
+        chunks = name.split("@")
+        if len(chunks) == 1:
+            return name, None
+        if chunks[1] == "" or len(chunks) > 2:
+            raise ModelError("Invalid model name '%s'" % name)
+        return chunks[0], chunks[1]
+
+    @staticmethod
+    def has_version_str(name):
+        '''Determines whether the given model name has a version string.'''
+        return bool(Model.parse_name(name)[1])
+
+    @classmethod
+    def from_dict(cls, d):
+        '''Constructs a Model from a JSON dictionary.'''
+        return cls(
+            d["base_name"], d["base_filename"],
+            ModelManager.from_dict(d["manager"]), d["date_created"],
+            d.get("version", None))
+
+
+class ModelWeights(object):
+    '''Base class for classes that encapsulate read-only model weights.
+
+    This class can load the model locally or from remote storage as needed.
+
+    Subclasses must implement the `_load` method to perform the actual loading.
+
+    Attributes:
+        model_name: the name of the model
+        model_path: the local path to the model on disk
+    '''
+
+    def __init__(self, model_name):
+        '''Initializes a ModelWeights instance.
+
+        Args:
+            model_name: the model to load
+
+        Raises:
+            ModelError: if the model was not found
+        '''
+        self.model_name = model_name
+        self.model_path = find_model(self.model_name)
+
+    def load(self):
+        '''Loads the model weights, downloading them from remote storage if
+        necessary.
+
+        Returns:
+            the model weights
+        '''
+        download_model(self.model_name, force=False)
+        return self._load()
+
+    def _load(self):
+        raise NotImplementedError("subclass must implement _load()")
+
+
+class NpzModelWeights(ModelWeights, dict):
```

Comment:
I get that these are to be viewed as ReadOnly but where/how are suitable files created?  By a different class/interface?  Any such model suffices?  This seems like one-half of the solution.

Reply:
What does suitable files mean? When one calls the `load` method of a `ModelWeights` class, it looks up the model `name` in the models manifest, automatically downloads it if necessary, loads the weights, and returns them to the user.

What's missing?

Reply:
I just mean the actual model files.  My read of the structure here is that this code awesome there exists other code for creating models and their files; then it is up to the developer to create this manifest information.  ...just seems like there is a disconnect here that the same class is not being used during the model creation as this ultimate model management/storage/etc.  (Or mixin.)  Something to think about; not critical.

Reply:
New models are created using `theta.core.models.publish_model`. It's a simple one-liner that sets everything up- uploads model to the cloud, logs it in a manifest, etc.

I had to split it up because there are certain proprietary aspects of our codebase that we don't want to put in ETA. We expose models to the world in ETA through a public Google Drive folder, but we can't let other people publish models to that location. And our code `theta.core.storage` code for reading/writing files to private Google Drive folders is not in ETA, so the aspects related to "publishing" THETA modules had to be split off into THETA.